### PR TITLE
feat: Add Ray head Pod high availability with Redis

### LINKED ISSUE #348: Ray Head pod High availability with Redis
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

Currently the JARK stack uses on-cluster, built-in redis cluster for RayServe. Best practice is to externalize redis for HA and resiliency of the inference endpoint.

#### Describe the solution you would like

Use MemoryDB for external redis cluster. Use auth mechanism to provide the Ray CR with a secret to connect to MemoryDB.

#### Describe alternatives you have considered

Elasticache Redis is also a possibility, need to pick the right service.

### PR DESCRIPTION
**Pull request by ratnopamc**
### What does this PR do?
Adds Ray Head pod High availability with enabling GCS fault tolerance
Adds terraform module to create an elastic cache Redis cluster in AWS
Updates website doc with a section on Ray head Pod high availability for Mistral-7b-inf2 blueprint.

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

### Motivation
By default, Ray head node is a single point of failure. If it crashes then Ray worker nodes get restarted. This introduces a downtime and is not desirable for RayServe applications. It's important to enable the GCS fault tolerance by connecting Ray head node to an external Redis cluster in order to provide high availability for the Ray head node and avoid restarting Ray workers in the event of a Ray head node crash.

Closes #348 

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [x] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: ai-ml/trainium-inferentia/outputs.tf [hunk 1]
```diff
@@ -2,3 +2,8 @@ output "configure_kubectl" {
   description = "Configure kubectl: make sure you're logged in with the correct AWS profile and run the following command to update your kubeconfig"
   value       = "aws eks --region ${var.region} update-kubeconfig --name ${local.name}"
 }
+
+output "elastic_cache_redis_cluster_arn" {
+  description = "Cluster arn of the cache cluster"
+  value       = length(module.elasticache) > 0 ? module.elasticache[0].cluster_arn : ""
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/outputs.tf:8]: change this to `try(module.elasticache[0].cluster_arn, null)`

File: gen-ai/inference/mistral-7b-rayserve-inf2/ray-service-mistral-ft.yaml [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  namespace: mistral
+data:
+  hf-token: $HUGGING_FACE_HUB_TOKEN
+---
+apiVersion: ray.io/v1
+kind: RayService
+metadata:
+  name: mistral
+  namespace: mistral
+  annotations:
```
Comment (vara-bonthu) [inline gen-ai/inference/mistral-7b-rayserve-inf2/ray-service-mistral-ft.yaml:22]: add a comment here about HA config

File: gen-ai/inference/mistral-7b-rayserve-inf2/ray-service-mistral-ft.yaml [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            - mountPath: /tmp/ray
+              name: ray-logs
+            resources:
+              limits:
+                cpu: "2"
+                memory: "20G"
+              requests:
+                cpu: "2"
+                memory: "20G"
+            env:
```
Comment (vara-bonthu) [inline gen-ai/inference/mistral-7b-rayserve-inf2/ray-service-mistral-ft.yaml:96]: add a comment for redis endpoint

File: gen-ai/inference/mistral-7b-rayserve-inf2/ray-service-mistral.yaml [hunk 4]
```diff
@@ -61,6 +61,7 @@ spec:
           namespace: mistral
       rayStartParams:
         dashboard-host: '0.0.0.0'
+        num-cpus: "0"
```
Comment (vara-bonthu) [inline gen-ai/inference/mistral-7b-rayserve-inf2/ray-service-mistral.yaml:64]: add a comment on why is this required

File: website/docs/gen-ai/inference/Mistral-7b-inf2.md [hunk 5]
```diff
@@ -231,6 +232,164 @@ Below screenshots provide some examples of the model response based on different
 
 ![Gradio Convo 2](img/mistral-conv-2.png)
 
+
+### Ray Head Node High Availability With Elastic Cache for Redis
+
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/Mistral-7b-inf2.md:237]: :::info

This Mistral7b deployment is using Ray Serve with High availability. If you have already deployed mistral7b in the previous steps, then you can delete the deployment and run the below steps.

:::

File: website/docs/gen-ai/inference/Mistral-7b-inf2.md [hunk 6]
```diff
@@ -231,6 +232,164 @@ Below screenshots provide some examples of the model response based on different
 
 ![Gradio Convo 2](img/mistral-conv-2.png)
 
+
+### Ray Head Node High Availability With Elastic Cache for Redis
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/Mistral-7b-inf2.md:236]: Move this to a new document under inference and call it as "RayServe High Availability."

File: website/docs/gen-ai/inference/Mistral-7b-inf2.md [hunk 7]
```diff
@@ -231,6 +232,164 @@ Below screenshots provide some examples of the model response based on different
...
+- A cache subnet group needs to be created at the time of creating the Redis cluster. A subnet group is a collection of subnets that you may want to designate for your caches in a VPC. ElastiCache uses that cache subnet group to assign IP addresses within that subnet to each cache node in the cache. The blueprint automatically adds all the subnets used by the EKS cluster in the subnet group for the Elastic cache Redis cluster.
+- Security Group - The Security Group assigned to the Redis cache needs to have an inbound rule that allows TCP traffic from EKS Cluster's worker node security group to the Redis cluster security group over port 6379. This is because the Ray head Pod needs to establish a connection to the Elastic cache Redis cluster over port 6379. The blueprint automatically sets up the security group with the inbound rule.
+
+To create the Redis cluster using Amazon Elastic Cache, please follow the below steps.
+
+First, enable the creation of the Redis cluster by setting the `enable_rayserve_ha_elastic_cache_redis` variable to `true` in `variables.tf` file. By default it's set to `false`.
+
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/Mistral-7b-inf2.md:261]: add pre-reqs here

**Review by vara-bonthu — COMMENTED** [review]
Thaks for the PR. Some minor comments

File: ai-ml/trainium-inferentia/elastic-cache-redis.tf [hunk 8]
```diff
@@ -0,0 +1,54 @@
+module "elasticache" {
+  count   = var.enable_rayserve_ha_elastic_cache_redis ? 1 : 0
```
Comment (bryantbiggs) [inline ai-ml/trainium-inferentia/elastic-cache-redis.tf:2]: all of our modules have a [`create`](https://github.com/terraform-aws-modules/terraform-aws-elasticache/blob/a483023ae9800c047d63206da7676afa79f1d82c/variables.tf#L1) variable to enable/disable resource creation. This is the correct way to toggle resource creation instead of using `count` on a module

```suggestion
  create   = var.enable_rayserve_ha_elastic_cache_redis
```

File: ai-ml/trainium-inferentia/outputs.tf [hunk 9]
```diff
@@ -2,3 +2,8 @@ output "configure_kubectl" {
   description = "Configure kubectl: make sure you're logged in with the correct AWS profile and run the following command to update your kubeconfig"
   value       = "aws eks --region ${var.region} update-kubeconfig --name ${local.name}"
 }
+
+output "elastic_cache_redis_cluster_arn" {
+  description = "Cluster arn of the cache cluster"
+  value       = try(module.elasticache[0].cluster_arn, null)
```
Comment (bryantbiggs) [inline ai-ml/trainium-inferentia/outputs.tf:8]: but using `create` properly, this simply becomes:

```suggestion
  value       = module.elasticache.cluster_arn
```

File: website/docs/gen-ai/inference/rayserve-ha.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+To address this, High Availability (HA) for the Ray head node is essential. Global Control Service (GCS) manages cluster-level metadata in a RayCluster. By default, the GCS lacks fault tolerance as it stores all data in-memory, and a failure can cause the entire Ray cluster to fail. To avoid this, one must add fault tolerance to Ray’s Global Control Store (GCS), which allows the Ray Serve application to serve traffic even when the head node crashes. In the event of a GCS restart, it retrieves all the data from the Redis instance and resumes its regular functions.
+
+![Ray-head-worker-redis](img/ray-head-ha-1.png)
+
+![Ray-head-ha](img/ray-head-ha-2.png)
+
+Following sections provide the steps on how to enable GCS fault tolerance and ensure high availability for the Ray head Pod. We're using the `Mistral-7B-Instruct-v0.2` model to demonstrate the Reay head high availability.
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/rayserve-ha.md:24]: correct the spelling suggested by workflow

File: website/docs/gen-ai/inference/rayserve-ha.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+1. [aws cli](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html)
+2. [kubectl](https://Kubernetes.io/docs/tasks/tools/)
+3. [terraform](https://learn.hashicorp.com/tutorials/terraform/install-cli)
+4. [envsubst](https://pypi.org/project/envsubst/)
+5. [jq](https://jqlang.github.io/jq/download/)
+
+First, enable the creation of the Redis cluster by setting the `enable_rayserve_ha_elastic_cache_redis` variable to `true` in `variables.tf` file. By default it's set to `false`.
+
+Then, run the `install.sh` script to install the EKS cluster with KubeRay operator and other add ons.
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/rayserve-ha.md:58]: may be change this to `add-ons`

File: ai-ml/trainium-inferentia/variables.tf [hunk 12]
```diff
@@ -140,3 +140,9 @@ variable "kms_key_admin_roles" {
   type        = list(string)
   default     = []
 }
+
+variable "enable_rayserve_ha_elastic_cache_redis" {
+  description = "Flag to enable Ray Head High Availability with Elastic Cache for Redis"
+  type        = bool
+  default     = true
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:147]: Let's make this `false` as default and enable it as a part of the RayHA deployment. Just to avoid deploying this cluster for other training blueprints

File: website/docs/gen-ai/inference/rayserve-ha.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+```bash
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/rayserve-ha.md:60]: add a step to enable `export TF_VAR_enable_rayserve_ha_elastic_cache_redis=true`

File: website/docs/gen-ai/inference/rayserve-ha.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+## Cleanup
+
+Finally, we'll provide instructions for cleaning up and deprovisioning the resources when they are no longer needed.
+
+**Step1:** Delete Gradio App and mistral Inference deployment
+
+```bash
+cd gen-ai/inference/mistral-7b-rayserve-inf2
+kubectl delete -f gradio-ui.yaml
+kubectl delete -f ray-service-mistral.yaml
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/rayserve-ha.md:199]: should this be `ray-service-mistral-ft.yaml`

Comment (ratnopamc) [reply website/docs/gen-ai/inference/rayserve-ha.md]: yes, changed.

**Review by vara-bonthu — APPROVED** [review]
@ratnopamc Nice work 👏🏼
